### ISSUE TITLE
Query: Raw SQL IN subqueries

### ISSUE THREAD

**Issue by benmccallum** [dotnet/efcore#21020]
Currently we are using SqlKata to do cursor-based paging, but with some other tickets in here (#17065), we'll potentially be able to unlock a similar story on EF Core natively. 

Unfortunately, our security (in GraphQL resolvers) will be probably the final blocker preventing us from getting across fully to EF Core. In a GraphQL world, if your client provides no arguments, you need to be able filter the results to only those they should be able to access (e.g. their own bookings), a.k.a. graph-based access control (GBAC). 

We facilitate this by building up a SQL subquery (using SqlKata to progressively build up a query that filters to their graph-based access) and then, since we are using SqlKata at the top to execute the query for results, we can just do `query.WhereIn("BookingId", subQueryForBookingIdsAllowed)`.

So the problem is, even if we can connection page with `SkipWhile` after #17065, we'll still be "Achilles heeled" by this.

Could we add `WhereIn` extension method/s off `IQueryable<T>` that just tack this on? 
* `public static IQueryable<T> WhereIn<T, TProperty>(this IQueryable<T> queryable, Expression<Func<T, TProperty>> property, string query)`
* `public static IQueryable<T> WhereIn<T, TProperty>(this IQueryable<T> queryable, string propertyName, string query)`

Also (I don't need it in my case, since I need an AND), but maybe to make doing `OR` easier (though it'd be possible with expression builder libraries) we'd want to also support something like `.Where(booking => booking.BookingId.In(query))`, though probably with a more explicit name than `In`.

**Comment by roji** [comment]
@benmccallum there are various scenarios in which EF Core already produces SQL IN with a subquery. For example:

```c#
var results = ctx.Blogs.Where(b => b.Posts.Select(d => d.Id).Contains(b.Id)).ToList();

public class Blog
{
    public int Id { get; set; }
    public string Name { get; set; }
    public List<Post> Posts { get; set; }
}

public class Post
{
    public int Id { get; set; }
    public string Title { get; set; }
}
```

Produces:

```sql
SELECT b."Id", b."DetailsId", b."Name"
FROM "Blogs" AS b
WHERE b."Id" IN (
    SELECT p."Id"
    FROM "Posts" AS p
    WHERE b."Id" = p."BlogId"
)
```

What we need is a more specific, detailed case which you're trying to handle.

**Comment by benmccallum** [comment]
Hi @roji , sorry I never got back to you on this.

I've sort of described a use case in the original issue description but I'll elaborate a bit more below.

Say I want to expose a list of blog posts from my GraphQL Query type and I have some filtering args like a, b, c.
```graphql
type Query {
  posts(a: Int, b: Date, c: String) : [Post!]!
}
```

In my GraphQL resolver I'd do something like:

```c#
var results = ctx.Posts.Where(p => /* Filter by a, b and c */).ToList();
```

Problem is in a GraphQL setting, since `posts` are exposed as a root field on my `Query` type, authorization to view posts becomes an issue. There's a few schools of thought on how this should work:
1. Filter posts to only ever return those that the viewer has access to
2. Don't filter, then authorize after the load and throw. e.g. Client needs to filter to what they'd have access to, e.g. by their user id.

Say I want to achieve approach number 1 and we have the following rules:
* an anonymous user to the blog website can't view posts that are unpublished, 
* an authenticated user that's the author of a blog post can see that post (e.g. to edit the draft)
* an authenticated user that's NOT the author, shouldn't see it, unless they are the editor in charge of the team that the author is in

We use SqlKata quite effectively to compose separate queries, e.g. add ids allowed by concating queryies with `UNION` or doing an AND/OR using a subquery formed from another query, etc. And we also compile these to SQL and verify their output in unit tests.

Long story short, it'd be cool if we could continue to use SqlKata to build this subquery, compile it to raw SQL, and then have a method off IQueryable for EF like `.WhereIn()`, we in example above we could do:
```c#
var results = ctx.Posts
    .Where(p => /* Filter by a, b and c */)
    .WhereIn(p => p.Id, rawSqlFromSqlKata)
    .ToList();
```

Potentially this composability of queries and the ability to test them could be done with LINQ expressions though. I never really explored it as we went down a SqlKata path to achieve cursor-based paging (we don't expose `IQueryable<T>` directly out of our graphql resolvers like Michael was demo'ing the other day).

**Comment by benmccallum** [comment]
That's my explicit scenario, but this could be something that's useful to other people. Without the other couple of issues around SkipWhile / cursor-based paging, and with the way we compose these authorization filter subqueries, currently we actually build a query in SqlKata that does the filtering, paging and sorting, fire that off to the DB for a list of the record ids. Then we fire another query via EF simply for those ids, ordered, to grab them back typed (actually pushing that through AutoMapper's `ProjectTo<T>` and passing it the fields requested by the graphql client to do an optimized projection). Things get a little more complicated if you don't pass IQueryable out of your GraphQL resolvers and let HotChocolate do the heavy lifting ;)

**Comment by roji** [comment]
@benmccallum specifically the above can achieved via EF's [raw SQL querying](https://docs.microsoft.com/en-us/ef/core/querying/raw-sql), which is LINQ-composable:

```c#
var results = ctx.Posts
    .FromSqlRaw($@"SELECT .... FROM [Posts] WHERE Id IN ({rawSqlFromSqlKata})")
    .Where(p => /* Filter by a, b and c */)
    // continue composing any LINQ operators as usual
    .ToList();
```

That's not a complete answer, since you raw SQL must be in the first part of the query, i.e. you can't start with LINQ and then add raw SQL. It also forces you to specify the SELECT and projection, so isn't ideal.

Otherwise, yeah, the "queryable" way would be for your filtering, paging and sorting to be generated via LINQ operators, rather than as a raw SQL - that's indeed the GraphQL approach.

**Comment by benmccallum** [comment]
I didn't know that you could chain FromSqlRaw with additional methods, so that's kinda cool, but it's still not enough to be able to do the projection like you said... Hmm

I need to have a think about how hard it would be for us to transition to HotChocolate's projection support but we've built a lot around the AutoMapper projection, mostly around having some properties rely on others so they are projected (like how Michael in his demo needed to have another resolver to force a FK I'd property so the relation resolver and its dataloader would be able to use it), so it's not super easy...

**Comment by Bouke** [comment]
I have a similar use-case. I'm composing an `IQueryable<T>` from a search, which is both defined by the user (dynamic search options) and by the system (e.g. access control). Both of these currently require the use of `FromSqlInterpolated`, but composing them is hard: I need to combine these predicates manually. Having something like `WhereSqlInterpolated(FormattableString)` would be very helpful to compose various criteria.

**Comment by roji** [comment]
This issue seems to be about allowing the integration of SQL fragments inside various places in the query, not at the query root level (e.g. an arbitrary WHERE clause). The main problem here is that SQL fragments aren't isolated: they need to refer e.g. to table aliases, which are managed by EF, uniquified, etc. So I think anything here would be quite brittle (e.g. EF changes its table alias uniquification algorithm).

Some alternatives:

* It's already possible to dynamically construct the desired fragment as an expression tree, and EF will translate that as usual; so this really is only useful when another component is used to generate SQL which then needs to be integrated in the EF query. 
* It's also possible to encapsulate the logic in a stored function and call that.
* To really integrate arbitrary SQL inside e.g. a WHERE clause, you can also integrate a dummy WHERE query, and use a command interceptor to find and replace it with what you want. You can even parse the dummy WHERE query to get the uniquified table alias to make the mechanism more resilient - this is better than having EF integrating a pre-baked SQL which assumes anything about table aliases.

So I'm not sure we should do this - opening to discuss whether we want to keep this on the backlog or close.

**Comment by benmccallum** [comment]
Thanks for the additional context Shay. I agree this isn't really feasible given the aliasing going on. Thanks for the other suggestions though, something for me to think about.
